{
  "url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257",
  "repository_url": "https://api.github.com/repos/quarkusio/quarkus",
  "labels_url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257/labels{/name}",
  "comments_url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257/comments",
  "events_url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257/events",
  "html_url": "https://github.com/quarkusio/quarkus/issues/25257",
  "id": 1220839390,
  "node_id": "I_kwDOCFbutM5IxIfe",
  "number": 25257,
  "title": "Quarkus 2.9.0.CR1: Async Scheduled methods have unsafe sub-context",
  "labels": [
    {
      "id": 985346620,
      "node_id": "MDU6TGFiZWw5ODUzNDY2MjA=",
      "url": "https://api.github.com/repos/quarkusio/quarkus/labels/kind/bug",
      "name": "kind/bug",
      "color": "d73a4a",
      "default": false,
      "description": "Something isn't working"
    },
    {
      "id": 1287515315,
      "node_id": "MDU6TGFiZWwxMjg3NTE1MzE1",
      "url": "https://api.github.com/repos/quarkusio/quarkus/labels/area/kotlin",
      "name": "area/kotlin",
      "color": "0366d6",
      "default": false,
      "description": ""
    },
    {
      "id": 1321590926,
      "node_id": "MDU6TGFiZWwxMzIxNTkwOTI2",
      "url": "https://api.github.com/repos/quarkusio/quarkus/labels/area/scheduler",
      "name": "area/scheduler",
      "color": "0366d6",
      "default": false,
      "description": ""
    }
  ],
  "state": "closed",
  "locked": false,
  "milestone": {
    "url": "https://api.github.com/repos/quarkusio/quarkus/milestones/185",
    "html_url": "https://github.com/quarkusio/quarkus/milestone/185",
    "labels_url": "https://api.github.com/repos/quarkusio/quarkus/milestones/185/labels",
    "id": 7939694,
    "node_id": "MI_kwDOCFbutM4AeSZu",
    "number": 185,
    "title": "2.9.0.Final",
    "description": "",
    "creator": {
      "login": "gsmet",
      "id": 1279749,
      "node_id": "MDQ6VXNlcjEyNzk3NDk=",
      "avatar_url": "https://avatars.githubusercontent.com/u/1279749?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/gsmet",
      "html_url": "https://github.com/gsmet",
      "followers_url": "https://api.github.com/users/gsmet/followers",
      "following_url": "https://api.github.com/users/gsmet/following{/other_user}",
      "gists_url": "https://api.github.com/users/gsmet/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/gsmet/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/gsmet/subscriptions",
      "organizations_url": "https://api.github.com/users/gsmet/orgs",
      "repos_url": "https://api.github.com/users/gsmet/repos",
      "events_url": "https://api.github.com/users/gsmet/events{/privacy}",
      "received_events_url": "https://api.github.com/users/gsmet/received_events",
      "type": "User",
      "site_admin": false
    },
    "open_issues": 0,
    "closed_issues": 21,
    "state": "closed",
    "created_at": "2022-05-03T17:27:43Z",
    "updated_at": "2022-05-18T08:23:57Z",
    "due_on": null,
    "closed_at": "2022-05-04T12:26:41Z"
  },
  "comments": 11,
  "created_at": "2022-04-29T13:34:46Z",
  "updated_at": "2022-05-03T17:30:15Z",
  "closed_at": "2022-04-29T16:43:07Z",
  "active_lock_reason": null,
  "body": "### Describe the bug\n\nWhen using an async `@Scheduled` method returning a void or using Kotlin coroutines, the provided sub-context seems to be unsafe and throws an error when e.g. using hibernate reactive with it. \r\n\r\nE.g. using a code like\r\n\r\n```\r\nclass TestScheduler(private val testDao: TestDao) {\r\n\r\n    @Scheduled(every = \"5s\")\r\n    fun every5s(): Uni<Void> {\r\n        return testDao.getEntity(UUID.randomUUID()).replaceWithVoid()\r\n    }\r\n}\r\n\r\n@ApplicationScoped\r\nclass TestDao(private val sessionFactory: Mutiny.SessionFactory) {\r\n    fun getEntity(id: UUID): Uni<TestEntity?> = sessionFactory.withTransaction { s, _ ->\r\n        s.find(TestEntity::class.java, id)\r\n    }\r\n}\r\n\r\n```\r\n\r\nends in a crash like\r\n\r\n```\r\n2022-04-29 15:32:05,016 ERROR [io.qua.sch.com.run.StatusEmitterInvoker] (vert.x-eventloop-thread-0) Error occured while executing task for trigger IntervalTrigger [id=1_org.acme.TestScheduler_ScheduledInvoker_every5s_c56c24f01b03a432de6e9b18de0f5a7dfbc05db8, interval=5000]: java.lang.IllegalStateException: The current operation requires a safe (isolated) Vert.x sub-context, but the current context hasn't been flagged as such. You can still use Hibernate Reactive, you just need to avoid using the methods which implicitly require accessing the stateful context, such as MutinySessionFactory#withTransaction and #withSession.\r\n\tat io.quarkus.vertx.core.runtime.context.VertxContextSafetyToggle.checkIsSafe(VertxContextSafetyToggle.java:80)\r\n\tat io.quarkus.vertx.core.runtime.context.VertxContextSafetyToggle.validateContextIfExists(VertxContextSafetyToggle.java:63)\r\n\tat io.quarkus.hibernate.reactive.runtime.customized.CheckingVertxContext.get(CheckingVertxContext.java:46)\r\n\tat org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.withSession(MutinySessionFactoryImpl.java:189)\r\n\tat org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.withTransaction(MutinySessionFactoryImpl.java:261)\r\n\tat org.acme.db.TestDao.getEntity(TestDao.kt:12)\r\n\tat org.acme.db.TestDao_Subclass.getEntity$$superforward1(Unknown Source)\r\n\tat org.acme.db.TestDao_Subclass$$function$$1.apply(Unknown Source)\r\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)\r\n\tat io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)\r\n```\r\n\n\n### Expected behavior\n\n@Scheduled should have a safe context that can be used with hibernate reactive.\n\n### Actual behavior\n\nIt crashes\n\n### How to Reproduce?\n\n1. Download reproducer project and unzip:\r\n[2022-04-29_scheduler-unsafe.zip](https://github.com/quarkusio/quarkus/files/8591125/2022-04-29_scheduler-unsafe.zip)\r\n2. run the included `docker-compose.yml` to set up a local postgres container. \r\n3. run with `./gradlew quarkusDev`\r\n4. check the log for the crash\n\n### Output of `uname -a` or `ver`\n\n_No response_\n\n### Output of `java -version`\n\n_No response_\n\n### GraalVM version (if different from Java)\n\n_No response_\n\n### Quarkus version or git rev\n\n2.9.0.CR1\n\n### Build tool (ie. output of `mvnw --version` or `gradlew --version`)\n\ngradle\n\n### Additional information\n\n_No response_",
  "reactions": {
    "url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257/reactions",
    "total_count": 0,
    "+1": 0,
    "-1": 0,
    "laugh": 0,
    "hooray": 0,
    "confused": 0,
    "heart": 0,
    "rocket": 0,
    "eyes": 0
  },
  "timeline_url": "https://api.github.com/repos/quarkusio/quarkus/issues/25257/timeline",
  "performed_via_github_app": null,
  "state_reason": "completed"
}
